Source/WebCore/page/unified-text-replacement/UnifiedTextReplacementController.mm

```diff
@@ -174,79 +183,75 @@
 
-        sessionRange = contextRangeForSessionWithID(session.identifier);
-        if (!sessionRange) {
-            ASSERT_NOT_REACHED();
-            return;
-        }
+        sessionRange = makeSimpleRange(state->contextRange);
 
         auto newRangeWithOffset = CharacterRange { locationWithOffset, replacementData.replacement.length() };
-        auto newResolvedRange = resolveCharacterRange(*sessionRange, newRangeWithOffset);
+        auto newResolvedRange = resolveCharacterRange(sessionRange, newRangeWithOffset);
 
         auto originalString = [context.attributedText.nsAttributedString() attributedSubstringFromRange:replacementData.originalRange];
 
         auto markerData = DocumentMarker::UnifiedTextReplacementData { originalString.string, replacementData.identifier, session.identifier, DocumentMarker::UnifiedTextReplacementData::State::Pending };
         addMarker(newResolvedRange, DocumentMarker::Type::UnifiedTextReplacement, markerData);
 
-        additionalOffset += static_cast<int>(replacementData.replacement.length()) - static_cast<int>(replacementData.originalRange.length);
+        state->replacementLocationOffset += static_cast<int>(replacementData.replacement.length()) - static_cast<int>(replacementData.originalRange.length);
     }
 
     if (finished)
-        document->selection().setSelection({ *sessionRange });
+        document->selection().setSelection({ sessionRange });
 }
 
-void UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement(const UnifiedTextReplacement::Session& session, UnifiedTextReplacement::Replacement::State state, const UnifiedTextReplacement::Replacement& replacement, const UnifiedTextReplacement::Context&)
+void UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement(const UnifiedTextReplacement::Session& session, UnifiedTextReplacement::Replacement::State newReplacementState, const UnifiedTextReplacement::Replacement& replacement, const UnifiedTextReplacement::Context&)
 {
-    RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement (%s) [new state: %hhu, replacement: %s]", session.identifier.toString().utf8().data(), enumToUnderlyingType(state), replacement.identifier.toString().utf8().data());
+    RELEASE_LOG(UnifiedTextReplacement, "UnifiedTextReplacementController::textReplacementSessionDidUpdateStateForReplacement (%s) [new state: %hhu, replacement: %s]", session.identifier.toString().utf8().data(), enumToUnderlyingType(newReplacementState), replacement.identifier.toString().utf8().data());
 
     RefPtr document = this->document();
     if (!document) {
         ASSERT_NOT_REACHED();
         return;
     }
 
-    auto sessionRange = contextRangeForSessionWithID(session.identifier);
-    if (!sessionRange) {
+    auto* state = stateForSession<UnifiedTextReplacement::Session::ReplacementType::PlainText>(session);
```

Comment:
Hm... it's probably okay as long as there's no reentrancy, but leaving raw pointers on the stack like this seems a bit counter to smart pointer deployment in WebKit.

Maybe we can use `CheckedPtr` instead?

Reply:
when I try that, I get
```
/Users/richardrobinson/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/CheckedPtr.h:179:18: error: no member named 'incrementPtrCount' in 'WebCore::UnifiedTextReplacementController::PlainTextState'
```

Reply:
oh, should I conform those types to `CanMakeCheckedPtr`?

Reply:
oh I guess I would also have to define a constructor, and also WTF_STRUCT_OVERRIDE_DELETE_FOR_CHECKED_PTR

```
    struct RichTextState : CanMakeCheckedPtr<RichTextState> {
        WTF_MAKE_STRUCT_FAST_ALLOCATED;
        WTF_STRUCT_OVERRIDE_DELETE_FOR_CHECKED_PTR(RichTextState);

        RichTextState(const Ref<Range>& contextRange, const Vector<Ref<ReplaceSelectionCommand>>& commands)
            : contextRange(contextRange),
              commands(commands)
        {
        }

        Ref<Range> contextRange;
        Vector<Ref<ReplaceSelectionCommand>> commands;
    };
```